lighthouse-logger/index.js

```diff
@@ -128,6 +128,72 @@ class Log {
     marky.stop(id);
   }
 
+  /* eslint-disable no-invalid-this */
+  static timeDecorate(originalFn, opts) {
+    return function timeDecoratedFn(...args) {
+      const timeStartLogLevel = opts.timeStartLogLevel || 'log';
+      const timeEndLogLevel = opts.timeStartLogLevel || 'verbose';
+
+      let msg;
+      if (typeof opts.msg === 'string') {
+        msg = opts.msg;
+      } else if (typeof opts.msg === 'function') {
+        msg = opts.msg.bind(this)(...args);
+      }
+      if (!msg) {
+        throw new Error('expected msg');
+      }
+
+      let id;
+      if (typeof opts.id === 'string') {
+        id = opts.id;
+      } else if (typeof opts.id === 'function') {
+        id = opts.id.bind(this)(...args);
+      } else {
+        id = `lh:${originalFn.name}`;
+      }
+      if (!id) {
+        throw new Error('expected id');
+      }
+
+      const status = {msg, id};
+      Log.time(status, timeStartLogLevel);
+
+      let result;
+      try {
+        result = originalFn.bind(this)(...args);
+      } catch (err) {
+        err.stack = err.stack.replace(/.* at timeDecoratedFn .*\n/g, '');
+        throw err;
+      }
+
+      if (result && typeof result.then === 'function') {
+        return result.then(value => {
+          Log.timeEnd(status, timeEndLogLevel);
+          return value;
+        }).catch(err => {
+          err.stack = err.stack.replace(/.* at timeDecoratedFn .*\n/, '');
```

Comment:
wouldn't work if minified. Is just the viewer minified? is it mangled? I need to check.

Reply:
Even if yes ... I don't think it matters.

Reply:
what about packages that bundle lighthouse (and likely minify) themselves? AFAICT, minifiers like `uglify-js` do not support "do not mangle me" comment markers, which would be the only way to ensure `timeDecoratedFn` is always named as such.

Anyway, it's not a big deal. Even if this fails, that just means an extra stack in an already unreadable (minified) stack trace. Unless source maps are applied, then it would show up. Should only address it if anyone ever complains.

Reply:
do we want to `timeEnd` in this `.catch` block too?

Reply:
yup!